### PR TITLE
Simplified the spack.util.gpg implementation

### PR DESCRIPTION
**Pull request by alalazo**
Modifications:
- [x] Removed a lot of complexity from classes + singleton + convenience functions
- [x] State is now recorded in 4 global variables that are not hidden in classes, API is unchanged
- [x] Fixed calls in a few tests accordingly
- [x] Added docstrings to functions

### CODE REVIEW

File: lib/spack/spack/util/gpg.py [hunk 1]
```diff
@@ -2,74 +2,123 @@
...
-    global _global_gpg_instance
+    # Set the value of GNUPGHOME to be used in this module
+    GNUPGHOME = (gnupghome or
+                 os.getenv('SPACK_GNUPGHOME') or
+                 spack.paths.gpg_path)
 
-    old_gnupg_home_override = _gnupg_home_override
-    old_global_gpg_instance = _global_gpg_instance
+    # Set the executable objects for "gpg" and "gpgconf"
+    GPG, GPGCONF = _gpg(), _gpgconf()
```
Comment (alalazo) [inline lib/spack/spack/util/gpg.py:63]: For reviewers. I started this refactoring to have a single place to modify (the line above) when adding bootstrapping for GnuPG. While I was at it I figured out I could simplify the implementation in this module without changing the API exposed to clients, so I took a stab at it.

**Comment by alalazo** [comment]
@opadron @scottwittenburg Do you have time to review this PR and tell me what you think? I ended up working on `spack.util.gpg` as part of a broader modification to enable `GnuPG` bootstrapping. Reading this module I figured out I could simplify a bit the implementation without modifying too much the external API, and so I took a stab at it and added docstrings.

File: lib/spack/spack/util/gpg.py [hunk 2]
```diff
@@ -2,74 +2,123 @@
...
+    GNUPGHOME = (gnupghome or
+                 os.getenv('SPACK_GNUPGHOME') or
+                 spack.paths.gpg_path)
+
+    # Set the executable objects for "gpg" and "gpgconf"
+    GPG, GPGCONF = _gpg(), _gpgconf()
+    GPG.add_default_env('GNUPGHOME', GNUPGHOME)
+    if GPGCONF:
+        GPGCONF.add_default_env('GNUPGHOME', GNUPGHOME)
+        # Set the socket dir is not using GnuPG defaults
```
Comment (opadron) [inline lib/spack/spack/util/gpg.py]: `s/is/if/`

**Review by opadron — DISMISSED** [review]
LGTM.  Thanks for the cleanup pass! :+1:

**Comment by alalazo** [comment]
@opadron @scottwittenburg Github is merciless in dismissing reviews, so I'm in need of another one :slightly_smiling_face:

**Review by scottwittenburg — APPROVED** [review]

**Review by tgamblin — APPROVED** [review]

File: lib/spack/spack/test/ci.py [hunk 3]
```diff
@@ -56,8 +56,6 @@ def test_urlencode_string():
     assert(s_enc == 'Spack+Test+Project')
 
 
-@pytest.mark.skipif(not spack.util.gpg.has_gpg(),
-                    reason='This test requires gpg')
```
Comment (tgamblin) [inline lib/spack/spack/test/ci.py:60]: I think we should put this back.  Or is GPG now mocked?

**Review by tgamblin — CHANGES_REQUESTED** [review]

**Review by tgamblin — APPROVED** [review]

**Review by becker33 — APPROVED** [review]

**Comment by alalazo** [comment]
Thanks all!

**Pull request merged**